lhotse/dataset/speech_recognition.py

```diff
@@ -35,21 +38,134 @@ def __init__(
         self.cuts = cuts.cuts
         self.root_dir = Path(root_dir) if root_dir else None
         self.cut_ids = list(self.cuts.keys())
+        if cuts[0].supervisions[0].graph_archive is not None:
+            self.train = True
+        else:
+            self.train = False
+
+    def parse_rxfile(file):
+        # separate offset from filename
+        if re.search(':[0-9]+$', file):
+            (file, offset) = file.rsplit(':', 1)
+        return file, int(offset)
 
     def __getitem__(self, idx: int) -> Dict[str, torch.Tensor]:
         cut_id = self.cut_ids[idx]
         cut = self.cuts[cut_id]
 
         features = torch.from_numpy(cut.load_features(root_dir=self.root_dir))
-
         # We assume there is only one supervision for each cut in ASR tasks (for now)
         if len(cut.supervisions) > 1:
-            logging.warning("More than one supervision in ASR task! Selected the first one and ignored others.")
+            logging.warning(
+                "More than one supervision in ASR task! Selected the first one and ignored others.")
+
+        if self.train:
+            graph_archive = cut.supervisions[0]
+            filename, offset = parse_rxfile(graph_archive)
+            fst = simplefst.StdVectorFst.read_ark(filename, offset)
+            graph = ChainGraph(fst)
 
         return {
             'features': features,
             'text': cut.supervisions[0].text
+            'graph': graph if self.train else None
         }
 
     def __len__(self) -> int:
         return len(self.cut_ids)
+
+
+""" These functions are copied from https://github.com/YiwenShaoStephen/pychain_example/blob/master/dataset.py
+    They do not perfectly fit with SpeechRecognitionDataset but would probably still be useful and necessary here
+    especially when people tend to load this SpeechRecognitionDataset with a pytorch Dataloader.
+    As the default pytorch dataloader can only batchify tensor-like 
+    (e.g. numpy array or pytorch tensor) items from the __getitem__ function but we have a dictionary output in
+    SpeechRecognitionDataset so that we need to give customized dataloader here (another option is to leave it to
+    the users themselves to decide how to load data). All these usages below are quite basic and flexible and can
+    be replaced/re-writtern by other functions. The only special thing here is to use ChainGraphBatch class to 
+    batchify a list of ChainGraph.
+"""
+
+
+def _collate_fn_train(batch):
```

Comment:
I am not very familiar with PyChain graphs design so I will need some more input from you to understand it better. First, let me share what my original plan is, and let's figure out if it makes sense to follow that and if that's possible to do with PyChain. 

I'd like the `CutSet` to be the only data structure where we "mutate" the data - i.e. trim, mix, pad, and so on. These operations should lead the user to one of the following scenarios:
1) all "cuts" are of the same duration (e.g. all cuts are 10 seconds long);
2) the "cuts" are in buckets of the same duration (e.g. cuts are exactly either 5s, 10s, or 20s long).

Then, the user passes that `CutSet` to a `*Dataset` class that only needs to call `load_features`/`load_audio` to get the inputs of the requested duration. It's also the `*Dataset` class that needs to take the supervisions present in a given cut and create a meaningful representation suitable for training.

I'd like the CutSet to be able to cut not only the features/audio, but also the supervisions, like the `ChainGraph`. It already adjusts the supervision time bounds. I'm wondering whether (and how) we can trim/pad PyChain graph supervisions. I think you're already doing that in the `ChainGraphBatch` class though. I'm thinking, it would be ideal if, during the padding/trimming of the `Cut`, we could insert the information about the graph offset and the number of graph states to be loaded in the supervision object. An example:

1) an utterance (`SupervisionSegment`) originally has 50 seconds (super long utterance)
2) we decided to cut a 20 seconds long cut starting at second 10 (creating `Cut`s with `SupervisionSegment` that is only 20s long)
3) while cutting, we put the information in the `PyChainGraph` inside the `SupervisionSegment` object to load the graph only from state 1000 until the state 3000 (assuming 1 state == 10ms)
4) in the `SpeechRecognitionDataset`, we call `Cut.load_features()` to get 20s of features, and `supervision.pychain_graph.load()` to get 20s of graph supervision

The collate_fn also gets simpler, as the only thing it needs to do is put all the graphs into a `ChainGraphBatch` (no padding, no trimming), and maybe it can delegate all the other work into the default collate_fn in PyTorch.